syntax = "proto3";

enum Type {
  INIT_REQUEST = 0;
  INIT_RESPONSE = 1;
  GET_ATTR_REQUEST = 2;
  GET_ATTR_RESPONSE = 3;
  OPEN_REQUEST = 4;
  OPEN_RESPONSE = 5;
  RELEASE_REQUEST = 6;
  RELEASE_RESPONSE = 7;
  READ_DIR_REQUEST = 8;
  READ_DIR_RESPONSE = 9;
  READ_REQUEST = 10;
  READ_RESPONSE = 11;
  WRITE_REQUEST = 12;
  WRITE_RESPONSE = 13;
  CREATE_REQUEST = 14;
  CREATE_RESPONSE = 15;
  MKDIR_REQUEST = 16;
  MKDIR_RESPONSE = 17;
  UNLINK_REQUEST = 18;
  UNLINK_RESPONSE = 19;
  RMDIR_REQUEST = 20;
  RMDIR_RESPONSE = 21;
  RENAME_REQUEST = 22;
  RENAME_RESPONSE = 23;
  CHMOD_REQUEST = 24;
  CHMOD_RESPONSE = 25;
  TRUNCATE_REQUEST = 26;
  TRUNCATE_RESPONSE = 27;
  MKNOD_REQUEST = 28;
  MKNOD_RESPONSE = 29;
  LINK_REQUEST = 30;
  LINK_RESPONSE = 31;
  SYMLINK_REQUEST = 32;
  SYMLINK_RESPONSE = 33;
  READ_LINK_REQUEST = 34;
  READ_LINK_RESPONSE = 35;
  STATFS_REQUEST = 36;
  STATFS_RESPONSE = 37;
  FSYNC_REQUEST = 38;
  FSYNC_RESPONSE = 39;
  SETXATTR_REQUEST = 40;
  SETXATTR_RESPONSE = 41;
  GETXATTR_REQUEST = 42;
  GETXATTR_RESPONSE = 43;
  LISTXATTR_REQUEST = 44;
  LISTXATTR_RESPONSE = 45;
  REMOVEXATTR_REQUEST = 46;
  REMOVEXATTR_RESPONSE = 47;
  OPENDIR_REQUEST = 48;
  OPENDIR_RESPONSE = 49;
  RELEASEDIR_REQUEST = 50;
  RELEASEDIR_RESPONSE = 51;
  FSYNCDIR_REQUEST = 52;
  FSYNCDIR_RESPONSE = 53;
  UTIMENS_REQUEST = 54;
  UTIMENS_RESPONSE = 55;
  ACCESS_REQUEST = 56;
  ACCESS_RESPONSE = 57;
  LOCK_REQUEST = 58;
  LOCK_RESPONSE = 59;
  FLOCK_REQUEST = 60;
  FLOCK_RESPONSE = 61;
  FALLOCATE_REQUEST = 62;
  FALLOCATE_RESPONSE = 63;
  LSEEK_REQUEST = 64;
  LSEEK_RESPONSE = 65;
  LSP_REQUEST = 66;
  LSP_RESPONSE = 67;
}

message InitRequest { string name = 1; }

message InitResponse { int32 error = 1; }

message GetAttrRequest { string path = 1; }

message GetAttrResponse {
  int32 error = 1;
  int32 mode = 2;
  int32 nlink = 3;
  int64 size = 4;
  int64 atime = 5;
  int64 mtime = 6;
  int64 ctime = 7;
  bool own = 8;
  bool gown = 9;
}

message OpenRequest {
  string path = 1;
  int32 flags = 2;
}

message OpenResponse {
  int32 error = 1;
  int32 fd = 2;
}

message ReleaseRequest { int32 fd = 1; }

message ReleaseResponse { int32 error = 1; }

message ReadDirRequest { int32 directory_descriptor = 1; }

message ReadDirResponse {
  int32 error = 1;
  repeated string names = 2;
}

message ReadRequest {
  int32 fd = 1;
  int64 offset = 2;
  int32 size = 3;
}

message ReadResponse {
  int32 error = 1;
  bytes data = 2;
}

message WriteRequest {
  int32 fd = 1;
  int64 offset = 2;
  bytes data = 3;
}

message WriteResponse {
  int32 error = 1;
  int32 size = 2;
}

message CreateRequest {
  string path = 1;
  int32 mode = 2;
}

message CreateResponse {
  int32 error = 1;
  int32 fd = 2;
}

message MkdirRequest {
  string path = 1;
  int32 mode = 2;
}

message MkdirResponse { int32 error = 1; }

message UnlinkRequest { string path = 1; }

message UnlinkResponse { int32 error = 1; }

message RmdirRequest { string path = 1; }

message RmdirResponse { int32 error = 1; }

message RenameRequest {
  string old_path = 1;
  string new_path = 2;
  int32 flags = 3;
}

message RenameResponse { int32 error = 1; }

message ChmodRequest {
  string path = 1;
  int32 mode = 2;
}

message ChmodResponse { int32 error = 1; }

message TruncateRequest {
  string path = 1;
  int64 size = 2;
}

message TruncateResponse { int32 error = 1; }

message MknodRequest {
  string path = 1;
  int32 mode = 2;
  int32 dev = 3;
}

message MknodResponse { int32 error = 1; }

message LinkRequest {
  string old_path = 1;
  string new_path = 2;
}

message LinkResponse { int32 error = 1; }

message SymlinkRequest {
  string old_path = 1;
  string new_path = 2;
}

message SymlinkResponse { int32 error = 1; }

message ReadLinkRequest { string path = 1; }

message ReadLinkResponse {
  int32 error = 1;
  string path = 2;
}

message StatfsRequest { string path = 1; }

message StatfsResponse {
  int32 error = 1;
  int64 frsize = 2;
  int64 blocks = 3;
  int64 bfree = 4;
  int64 bavail = 5;
  int64 bsize = 6;
  int64 files = 7;
  int64 ffree = 8;
  int64 namemax = 9;
  int32 type = 10;
}

message FsyncRequest { int32 fd = 1; }

message FsyncResponse { int32 error = 1; }

message SetxattrRequest {
  string path = 1;
  string name = 2;
  bytes value = 3;
  int32 flags = 4;
}

message SetxattrResponse { int32 error = 1; }

message GetxattrRequest {
  string path = 1;
  string name = 2;
}

message GetxattrResponse {
  int32 error = 1;
  bytes value = 2;
}

message ListxattrRequest { string path = 1; }

message ListxattrResponse {
  int32 error = 1;
  repeated string names = 2;
}

message RemovexattrRequest {
  string path = 1;
  string name = 2;
}

message RemovexattrResponse { int32 error = 1; }

message OpendirRequest { string path = 1; }

message OpendirResponse {
  int32 error = 1;
  int32 directory_descriptor = 2;
}

message ReleasedirRequest { int32 directory_descriptor = 1; }

message ReleasedirResponse { int32 error = 1; }

message FsyncdirRequest { int32 directory_descriptor = 1; }

message FsyncdirResponse { int32 error = 1; }

message UtimensRequest {
  string path = 1;
  int64 atime = 2;
  int64 mtime = 3;
}

message UtimensResponse { int32 error = 1; }

message AccessRequest {
  string path = 1;
  int32 mode = 2;
}

message AccessResponse { int32 error = 1; }

message Lock {
  int32 l_type = 3;
  int32 l_whence = 4;
  int64 l_start = 5;
  int64 l_len = 6;
}

message LockRequest {
  int32 fd = 1;
  int32 cmd = 2;
  Lock lock = 3;
}

message LockResponse {
  int32 error = 1;
  optional Lock lock = 2;
}

message FlockRequest {
  int32 fd = 1;
  int32 op = 2;
}

message FlockResponse { int32 error = 1; }

message FallocateRequest {
  int32 fd = 1;
  int32 mode = 2;
  int64 offset = 3;
  int64 len = 4;
}

message FallocateResponse { int32 error = 1; }

message LseekRequest {
  int32 fd = 1;
  int64 offset = 2;
  int32 whence = 3;
}

message LseekResponse {
  int32 error = 1;
  int64 offset = 2;
}

message LspRequest {
  string payload = 1;
  string language = 2;
}

message LspResponse {
  string payload = 1;
  string language = 2;
}

message TeaConfigFile { map<string, string> language_configs = 1; }
